### PR TITLE
subsys: fs: implement support for external file systems

### PR DESCRIPTION
**Pull request by pabigot**
The documentation claims that Zephyr supports external file system implementations, and there's no reason not to do so.  However the implementation uses a fixed enumeration to index a map from file system types to implementations, which is not extensible.  Rework the API to allow this.
    
Note that the file system type cannot legally be an enum anymore, since we need to support file system types that don't have an identifier assigned in that enum.  Rely on the implicit conversion of enum values to int to preserve backwards compatibility.

Also make the file system implementation objects `const`.

This work inspired by #25570.

### CODE REVIEW

**Comment by pabigot** [comment]
This needs to be updated to select a maximum file name over all enabled file systems, which would conflict with #25852 so wait until that's merged.

**Comment by pabigot** [comment]
Added patch to allow configuring file name length when using external file systems, and unblocked.

**Comment by pabigot** [comment]
Added a couple more filesystem maintainer/collaborators to get reviews.

File: subsys/fs/Kconfig
Comment (nashif) [inline subsys/fs/Kconfig:31]: why are those being added here? they already exist in the same file

**Review by nashif — CHANGES_REQUESTED** [review]
Kconfig seems to be duplicated.
please fix the commit message, no need for "subsys/" in the commit header.

File: subsys/fs/Kconfig
Comment (pabigot) [reply subsys/fs/Kconfig:31]: Rebase conflict didn't get resolved correctly.

**Review by nashif — APPROVED** [review]

**Review by vanwinkeljan — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by nvlsianpu** [comment]
Nice job, would be great to see a usage example.

**Pull request merged**
